### ISSUE TITLE
BUG: np.interp(nan, x, y) doesn't return nan when len(x)=1

### ISSUE THREAD

**Issue by syrte** [numpy/numpy#13555]
### Reproducing code example:

```python
import numpy as np
np.interp(np.nan, [1], [2])
```
returns `2.0` while `nan` is expected.

Related issue: https://github.com/numpy/numpy/issues/605

**Comment by eric-wieser** [comment]
I'd argue this is correct. `y = np.interp(x, [1], [2])` is the function `y = 2`, which gives `y = 2` even if `x` is `nan`.

However, the behavior of `np.interp(np.nan, [0], [0], left=-1, right=1)` is definitely wrong, which is the step function, where `nan` should definitely give `nan`

**Comment by syrte** [comment]
As I see, `nan -> nan` would bring less surprise to users, and we can get a consistent result when specifying `left` and `right` (as in your example). Otherwise, we have to decide `nan` is on the left or right side.
 
`y = np.interp(x, [1], [2])` is trying to infer a function from a single point, it's hard to say it's really the function `y=2`.

**Comment by eric-wieser** [comment]
> it's hard to say it's really the function y=2.

It's not, because of the default behavior of `left` and `right` to take on the value at the first and last sample, respectively.

**Comment by syrte** [comment]
Somehow I would feel surprised, if `np.interp(x, [1], [2])` gives `2` only when both `left` and `right` are specified and equal, in any other cases the value is undefined (thus `nan`). 
So it seems much simpler to return `nan` for all cases.

**Comment by seberg** [comment]
@eric-wieser, I agree with the NaN argument for constant functions. On the other hand, it means that it is well defined exactly when the function is constant. And NaN is not a number, not "possibly any number" (honestly not sure how to think of it in this context)? But if you have `np.interp(np.nan, [1, 2], [2, 2])` the function is also constant, and it will propagate the NaN. Another strange corner case maybe:
```
np.interp(2.5, [1, 2, np.nan], [2, 2, 2])
```
(since `np.nan` is sorted to the end IIRC).

**Comment by eric-wieser** [comment]
"possibly any number" is the interpretation of NaN I've used before, and it always seems to give the right intuition.

**Comment by seberg** [comment]
Hmmm, `sqrt(-1) == NaN` (in the real domain) may be a bit funny in that sense. But that is not what I have squirms with. I do somewhat agree that it is weird that `intp` behaves different for a support of 1 compared to >1 (when the function is constant).

**Comment by mhvk** [comment]
Given that we chose to return `NaN` for the `>1` case (see #4302), it seems most logical to extend that to all cases.

**Comment by eric-wieser** [comment]
@mhvk: Perhaps it's not worth the cost of detecting it, but I'd argue that `interp(np.nan, [1, 1, 1, 1], [2, 2, 2, 2])` should return `2`.

**Comment by seberg** [comment]
Well, I think we should decide one of the two:
 1. Return always `np.nan` here.
 2. Detect any case of constant functions.

And frankly, my gut feeling is that 2. is not very practical? I agree that mathematically speeking 1. seems better.

**Comment by mhvk** [comment]
@eric-wieser - your example makes sense mathematically, but like @seberg it would seem special-casing the constant `yp` array case is not worth the trouble. It also would likely end up hiding bugs since `nan` in the `x` array would likely be because something went wrong earlier.

Maybe we should just think of the interpolation as always having a part `slope * x` - then, even for zero slope you get out `0*nan=nan` (which is not unreasonable: if not a number equals "any number", it can also be infinity).

**Comment by eric-wieser** [comment]
> Maybe we should just think of the interpolation as always having a part slope * x

Note that removing this assumption was how I was able to improve the behavior on samples containing infinity

**Comment by mhvk** [comment]
I looked at the [implementation](https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/compiled_base.c#L475) - and indeed it has a nice way to avoid calculating the slope unless necessary. It is also clear that why `NaN` does not propagate for `len(xp) == 1` - that situation is special-cased.

I guess in principle, one could do the check whether all `yp` are equal only once a first `NaN` is encountered, so that one gets consistent behaviour even though not influencing the speed for the case where no `NaN` are present. That said, for infinities, I think the expectation is clear, while for `NaN` it is much less so. I also don't like that one then starts to rely on equality of floats, which is always a bit hairy. For instance, imagine you're interpolating in `yp=exp(-x)` - if those `x` are very large, all yp are zero and you thus get 0 for each nan, while if they are just not large enough - perhaps because you switched to double precision - you get `NaN`.

**Comment by WarrenWeckesser** [comment]
There is precedent for @eric-wieser's point of view in the IEEE 754 specification itself.  I'm not an expert in IEEE 754 specification, so I'll quote this comment by Mark Dickinson (from a comment in in a discussion of a bug report on bugs.python.org):

> [...] for the same sort of reasons that IEEE 754 specifies that hypot(nan, inf) is inf rather than nan. Similarly, sumSquare and sumAbs from IEEE 754-2008 specify that infinities take precedence over NaNs, on the basis that the result doesn't change if the nan is replaced with any non-nan value.

Indeed,

```
In [32]: np.hypot(np.nan, np.inf)
Out[32]: inf
```

**Comment by WarrenWeckesser** [comment]
(On the other hand, if you apply the principle of replacing the nan with any non-nan value to see if the result is always the same, shouldn't `nan <= inf` be True?)
